### PR TITLE
add sessionId to java client lib request header

### PR DESCRIPTION
**Pull request by kawcl**

### CODE REVIEW

**Comment by kawcl** [comment]
I did not add the sessionId to the constructor of the `ExperimentClient` so not to break any current implementations using it. Also because it's an optional field.  Instead I provided a setter if a user wanted to optionally set it.  If it's thought to be better to include it with the constructor that can be added.

**Comment by jerith666** [comment]
You need to rev the version in the pom file.

Since clients will also need to rev the version in their pom files, I think it's fine to make breaking API changes like adding a c'tor argument.

I'd rather see the session ID as a required c'tor argument, since we believe it's best practice to provide one.  The caller can always fill it in with `UUID.randomUUID().toString()` if their environment doesn't provide anything better.  :)

**Comment by kawcl** [comment]
@jerith666 you want the setter completely removed with the constructor update?

File: clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java [hunk 1]
```diff
@@ -50,13 +50,12 @@
      *            Properties to permit users to control how the underlying JAX-RS
      *            client behaves. These are passed through to
      *            {@link javax.ws.rs.core.Configurable#property(String, Object)}. */
-	public ExperimentClient(String userId, String authToken, String baseUrl, Map<String, Object> properties) {
+	public ExperimentClient(String userId, String authToken, String sessionId, String baseUrl, Map<String, Object> properties) {
```
Comment (VivekFitkariwala) [inline clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:53]: In Javascript client we have kept sessionId optional so if it is not provided explicitly we can generate a random one.

**Review by jerith666 — APPROVED** [review]
This looks good to me -- passing the session ID in to the `APIService` ctor is a better approach.

I prefer requiring the argument for the reasons I gave earlier, but if others feel strongly about it, I'm okay with a second ctor that doesn't require it.

File: clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java (hunk 1 above)
Comment (kawcl) [reply clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:66]: I added in a second constructor without the session id argument, making it optional.  If that one is used then we will generate a random one

File: clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java [hunk 2]
```diff
@@ -55,8 +56,19 @@ public ExperimentClient(String userId, String authToken, String baseUrl, Map<Str
 			throw new IllegalArgumentException(INVALID_STUDENT_ID);
 		}
 		this.userId = userId;
+		this.apiService = new APIService(baseUrl, authToken, UUID.randomUUID().toString(), properties);
```
Comment (jerith666) [inline clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:59]: you don't need to copy-paste the whole body of the constructor.  Just do:

```
    this(userId, authToken, UUID.randomUUID().toString(), baseUrl, properties);
```

**Review by jerith666 — CHANGES_REQUESTED** [review]

File: clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java (hunk 2 above)
Comment (kawcl) [reply clientlibs/java/src/main/java/org/upgradeplatform/client/ExperimentClient.java]: good catch, it's been updated

**Review by VivekFitkariwala — APPROVED** [review]

**Review by jerith666 — APPROVED** [review]

**Comment by jerith666** [comment]
Looks good now, thanks!

**Pull request merged**